### PR TITLE
build: Initial handling for ARM64 windows compilation

### PR DESCRIPTION
**Pull request by cosmo0920**
In ARM64 Windows, winpthread, LuaJIT, and windows_exporter_metrics does not build and run correctly.

This PR makes initial handling for ARM64 Windows building.

* Disable LuaJIT building on ARM64 Windows by default
* Adjust control register pointing on ARM64 Windows for winpthread
* Fix incorrect double value returning on windows_exporter_metrics

For building ARM64 Windows, it can be:

```cmd
cmd> "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\Tools\VsDevCmd.bat" -arch=arm64 -host_arch=x64
cmd> cmake .. -DOPENSSL_ROOT_DIR="S:\dir\to\openssl-win-aarch64\OpenSSL" -DCMAKE_SYSTEM_NAME=Windows -DCMAKE_SYSTEM_VERSION=10.0 -DCMAKE_SYSTEM_PROCESSOR=ARM64 -G "NMake Makefiles"
```

`-DCMAKE_SYSTEM_NAME=Windows -DCMAKE_SYSTEM_VERSION=10.0 -DCMAKE_SYSTEM_PROCESSOR=ARM64` are needed for configuring ARM64 building with cmake.

```cmd
cmd> dumpbin /headers bin\fluent-bit.exe
PE signature found

File Type: EXECUTABLE IMAGE

FILE HEADER VALUES
            AA64 machine (ARM64)
               6 number of sections
        63A95331 time date stamp Mon Dec 26 16:54:25 2022
               0 file pointer to symbol table
               0 number of symbols
              F0 size of optional header
              22 characteristics
                   Executable
                   Application can handle large (>2GB) addresses
<snip>
```

Note that this is initial work for native ARM64 Windows building. Some of components do not work correctly such as LuaJIT. LuaJIT does not work ARM64 Windows for now: https://github.com/LuaJIT/LuaJIT/issues/593

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Example configuration file for the change
- [ ] Debug log output from testing the change

- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Attached [local packaging test](./packaging/local-build-all.sh) output showing all targets (including any new ones) build.

**Documentation**

- [ ] Documentation required for this feature

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by patrick-stephens** [comment]
@cosmo0920 I think this is good to get in but we need to be verifying these builds in CI somehow, having un-built targets otherwise will always mean errors can creep in. We can then also verify things with the `ok-package-test` label across all targets easily once #6687 is merged.

We're going to be looking at deploying self-hosted runners to speed things up in this area and support targets that hosted Github runners cannot - do you know if Windows ARM64 is one such target?

**Comment by cosmo0920** [comment]
In my Windows x64 box, the following steps are required to build ARM64 Windows binary:

0. prerequisites
 * Install ARM64 VC compiler components

1. enable ARM64 cross compiler

```cmd
cmd> "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\Tools\VsDevCmd.bat" -arch=arm64 -host_arch=x64           
```

2. Build ARM64 version of OpenSSL artifacts:

```powershell
PS (administrator)> choco install -y StrawberryPerl
```

3. Download and build OpenSSL

```cmd
cmd> git clone git@github.com:openssl/openssl.git
cmd> git checkout openssl-3.0.7
cmd> C:\Strawberry\perl\bin\perl.exe Configure VC-WIN64-ARM
cmd> nmake && nmake install DESTDIR=C:\path\to\OpenSSL-ARM64
```

4. Build fluent-bit with ARM64 cross compiler

```cmd
cmd> git clone git@github.com:fluent/fluent-bit.git
cmd> mkdir build_arm64
cmd> cd build_arm64
cmd>  cmake .. -DCMAKE_SYSTEM_NAME=Windows -DCMAKE_SYSTEM_VERSION=10.0 -DCMAKE_SYSTEM_PROCESSOR=ARM64 -G "NMake Makefiles" -DOPENSSL_ROOT_DIR="C:\path\to\OpenSSL-ARM64\Program Files\OpenSSL"
cmd> cmake --build .
```

Finally, I got a AA64 signature fluent-bit.exe:

```cmd
cmd> dumpbin /headers bin/fluent-bit.exe

Microsoft (R) COFF/PE Dumper Version 14.34.31937.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file bin\fluent-bit.exe

PE signature found

File Type: EXECUTABLE IMAGE

FILE HEADER VALUES
            AA64 machine (ARM64)
               6 number of sections
        63C667F1 time date stamp Tue Jan 17 18:18:41 2023
               0 file pointer to symbol table
               0 number of symbols
              F0 size of optional header
              22 characteristics
                   Executable
                   Application can handle large (>2GB) addresses

<snip>
```

**Comment by cosmo0920** [comment]
But this procedure is too early and rough for our CI tasks.
I'll try to create vcpkg version of procedure for ARM64 Windows architecture.

**Comment by cosmo0920** [comment]
The latest commit can handle ARM64 cross compilation on GitHub actions:

```
 -- The C compiler identification is MSVC 19.34.31937.0
-- The CXX compiler identification is MSVC 19.34.31937.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.34.31933/bin/Hostx64/arm64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.34.31933/bin/Hostx64/arm64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
```

ref: https://github.com/cosmo0920/fluent-bit/actions/runs/3983242614/jobs/6828396337#step:7:60

**Comment by cosmo0920** [comment]
And also, `ilammy/msvc-dev-cmd@v1` can handle ordinary setting up and cross compilation one in the same way instead of using `egor-tensin/vs-shell@v2`. This would be nice to use building for unified way to build x64 , x86 and cross arm64 binaries.

**Comment by cosmo0920** [comment]
@patrick-stephens I got succeeded to build with a cross compiler for Windows ARM64 on ordinary workers of GitHub Actions. Are there any blockers to merge this PR?

File: cmake/windows-setup.cmake
Comment (patrick-stephens) [inline cmake/windows-setup.cmake:10]: Probably should link out to why if there is a good URL for an issue, etc. so we can enable if/when LUA is supported. We probably need to document this as well, presumably this means no LUA filter or does it just reduce functionality in another way?

File: plugins/in_windows_exporter_metrics/we_util.c
Comment (patrick-stephens) [inline plugins/in_windows_exporter_metrics/we_util.c:61]: We were using 0 as an error return before but now a success return so want to check this is correct? Should we have enums for this?

Comment (cosmo0920) [reply plugins/in_windows_exporter_metrics/we_util.c:61]: Good point. Using `FLB_TRUE` or `FLB_FALSE` would be better.

File: cmake/windows-setup.cmake
Comment (cosmo0920) [reply cmake/windows-setup.cmake:12]: There is a meta issue for supporting LuaJIT on ARM64 Windows: https://github.com/fluent/fluent-bit/pull/6621/commits/186ec58892ad8fc80b6a47f87fccb9e2fee1111d
For ARM64 Windows, we should disable for now and will revisit to enable this functionality.

File: plugins/in_windows_exporter_metrics/we_util.c
Comment (cosmo0920) [reply plugins/in_windows_exporter_metrics/we_util.c]: I use `FLB_TRUE` and `FLB_FALSE` for returning values here. This should be cleaner whether error or not.
Fixed in https://github.com/fluent/fluent-bit/pull/6621/commits/5c5e5f16d976845d883ab382020c936de30dcf83.

File: cmake/windows-setup.cmake
Comment (patrick-stephens) [inline cmake/windows-setup.cmake:10]:
```suggestion
  # LuaJIT does not currently support Windows ARM64 architecture so we disable it for now.
```

**Review by patrick-stephens — COMMENTED** [review]
Am I going mad or did this use to have the CI workflow updates in to also build on the ARM64 target?

File: cmake/windows-setup.cmake
Comment (cosmo0920) [reply cmake/windows-setup.cmake]: Applied. Thanks!

**Comment by cosmo0920** [comment]
> Am I going mad or did this use to have the CI workflow updates in to also build on the ARM64 target?

CI will reflect this change after merging this PR. This is because our CI calls master version of workflows: 
https://github.com/fluent/fluent-bit/blob/master/.github/workflows/pr-windows-build.yaml#L21

When changing branch specific ones, I confirmed that Windows ARM64 target should be built: https://github.com/cosmo0920/fluent-bit/actions/runs/3983242614 (You can see Windows 64bit (Arm64) target there).

**Comment by patrick-stephens** [comment]
> CI will reflect this change after merging this PR. This is because our CI calls master version of workflows:
> https://github.com/fluent/fluent-bit/blob/master/.github/workflows/pr-windows-build.yaml#L21

Ah, we can change this now - it used to be a restriction on Github Actions but no longer. We can call a branch one now, just need to remove the branch lock e.g. https://github.com/fluent/fluent-bit/blob/16eae107868bf03da2fce6ca4acd3f6ab2d3c4ff/.github/workflows/build-branch-containers.yaml#L11

**Comment by cosmo0920** [comment]
Now, I removed the branch lock for pr-build-windows workflow. I'm waiting the CI task for cross compilation of ARM64 Windows.

**Comment by cosmo0920** [comment]
Now, winarm64 artifacts are uploaded as windows packages on GitHub Actions' summary page.

**Review by patrick-stephens — APPROVED** [review]

**Comment by edsiper** [comment]
for Monkey changes please do them first in monkey/monkey, once merged we update the lib/monkey directory

**Comment by cosmo0920** [comment]
Now, I rebased off master. Waiting for CI results....

**Comment by cosmo0920** [comment]
Hadolint task might be broken for their unintended change but not related this PR.
Other CI tasks seems to be fine.

**Comment by patrick-stephens** [comment]
> Hadolint task might be broken for their unintended change but not related this PR.
> Other CI tasks seems to be fine.

Looks like it is resolved now, all green

**Comment by edsiper** [comment]
FYI: PR has been squashed due to conflicts fixed

**Pull request merged**
